### PR TITLE
Bowser's Big Blast

### PR DESCRIPTION
**Pull request by CSharpM7**
# Bowser's Big Blast

Assets: [hdr-bowser-assets.zip](https://github.com/HDR-Development/HewDraw-Remix/files/12044058/hdr-bowser-assets.zip)
Video: https://twitter.com/CSharpM7/status/1672033862820728832  
This adds Project +'s Fireball to Bowser, alongside giving him Melee Nair and a new animation for Dash Attack.

## Dash Attack

https://github.com/HDR-Development/HewDraw-Remix/assets/13909643/e632d09d-80ec-4507-bac5-3d7e82807bfe

```diff
! New Animation: This animation greatly increases the range of Dash Attack, at the cost of having several less active frames 
+ Speed Multipler: 0.8->1.0
- Active Frames: 11-21 -> 12-15
+ Range increased
+ FAF: 47->44
! No longer has a late hitbox, but instead has a sourspot on his body and inner arm
+ Damage: 12/8(late) -> 12/10(sour)
+ BKB: 80 -> 80/85(sour)
```

## Neutral Air

nair

```diff
! Replaced with Melee Nair's animation
! Startup: 8
! Active Frames: 8-13,14-25
! Damage: 13.0/9.0
! Angle: 61
! KBG: 75/75
! BKB: 40/20
! FAF: 38
```

## Neutral Special: Fireball

https://github.com/HDR-Development/HewDraw-Remix/assets/13909643/c3c73a82-3d5b-4e7c-9b43-8ab815710506

fireball

```diff
Bowser now has access to P+'s Fireball move. This move takes 15 seconds to charge, and the charge is lost upon death. During training mode, Bowser will always start with a full charge. When you start to see Bowser's mouth glow red, you can activate this move with Neutral Special, launching a powerful ball of fire toward your opponents.

! Fireball Spawn Frame: 24
! FAF: 60

! Lifespan: 60 frames
! Fireball damage reduction after half of its lifespan
! Despawns upon hitting an opponent
! Damage: 10.8/7.2
! Angle: 55
! KBG: 60
! BKB: 80
```

## Neutral Special (During Cooldown): Flamebreath

```diff
- Using this move will halt Fireball's cooldown. For each frame during the main attack, 1 frame is added to the cooldown timer. Cooldown timer won't exceed its initial max time.
- Max distance and size halved
+ FAF (From cancel): 31->16
```

### CODE REVIEW

File: fighters/koopa/src/status.rs [hunk 1]
```diff
@@ -31,4 +54,195 @@ pub unsafe fn exec_special_hi_a(fighter: &mut L2CFighterCommon) -> L2CValue {
...
+    else{
+        fighter.sub_change_motion_by_situation(Hash40::new("special_n_max").into(), Hash40::new("special_air_n_max").into(), false.into());
+        fighter.sub_set_ground_correct_by_situation(true.into());
+
+        fighter.sub_shift_status_main(L2CValue::Ptr(specialnmax_main_loop as *const () as _))
+    }
+}
+
+unsafe extern "C" fn specialnmax_main_loop(fighter: &mut L2CFighterCommon) -> L2CValue {
+    fighter.sub_change_motion_by_situation(Hash40::new("special_n_max").into(), Hash40::new("special_air_n_max").into(), true.into());
```
Comment (jobrien97) [inline fighters/koopa/src/status.rs]: This should go behind an `StatusModule::is_situation_changed` check, so it's not running every frame

File: fighters/koopa/src/status.rs [hunk 2]
```diff
@@ -31,4 +54,195 @@ pub unsafe fn exec_special_hi_a(fighter: &mut L2CFighterCommon) -> L2CValue {
...
+        return 0.into();
+    }
+    if fighter.is_situation(*SITUATION_KIND_GROUND){
+        KineticModule::change_kinetic(fighter.module_accessor, *FIGHTER_KINETIC_TYPE_GROUND_STOP);
+        GroundModule::set_correct(fighter.module_accessor, GroundCorrectKind(*GROUND_CORRECT_KIND_GROUND_CLIFF_STOP_ATTACK));
+    }
+    else {
+        KineticModule::change_kinetic(fighter.module_accessor, *FIGHTER_KINETIC_TYPE_AIR_STOP);
+        GroundModule::set_correct(fighter.module_accessor, GroundCorrectKind(*GROUND_CORRECT_KIND_AIR));
+    }
```
Comment (jobrien97) [inline fighters/koopa/src/status.rs]: Should also go behind an `StatusModule::is_situation_changed` check; resetting kinetic type every frame can have strange side effects. This block would also need to go in the pre-loop main func once that is done.

File: fighters/koopa/src/status.rs [hunk 3]
```diff
@@ -31,4 +54,195 @@ pub unsafe fn exec_special_hi_a(fighter: &mut L2CFighterCommon) -> L2CValue {
...
+            1.0,
+            false,
+            0.0,
+            false,
+            false
+        );
+        let param_life =  WorkModule::get_param_float(weapon.module_accessor, hash40("param_breath"), hash40("life")) as i32;
+        WorkModule::set_int(weapon.module_accessor, param_life, *WEAPON_INSTANCE_WORK_ID_INT_LIFE);
+        WorkModule::set_int(weapon.module_accessor, param_life, *WEAPON_INSTANCE_WORK_ID_INT_INIT_LIFE);
+        println!("Fireball life: {param_life}");
```
Comment (jobrien97) [inline fighters/koopa/src/status.rs]: println :^)

File: romfs/source/fighter/koopa/motion/body/motion_patch.yaml [hunk 4]
```diff
@@ -50,10 +50,64 @@ attack_air_f:
     cancel_frame: 38
 attack_air_n:
   extra:
-    cancel_frame: 44
+    cancel_frame: 38
```
Comment (jobrien97) [inline romfs/source/fighter/koopa/motion/body/motion_patch.yaml:57]: Is this f38 ingame? You motion rated 4f away from the startup, so just making sure

**Review by jobrien97 — CHANGES_REQUESTED** [review]

File: romfs/source/fighter/koopa/motion/body/motion_patch.yaml (hunk 4 above)
Comment (CSharpM7) [reply romfs/source/fighter/koopa/motion/body/motion_patch.yaml:57]: Was able to verify this is Frame 38 in game

File: fighters/koopa/src/status.rs (hunk 2 above)
Comment (CSharpM7) [reply fighters/koopa/src/status.rs]: Added this to `main`, though if the `pre` status uses ``SITUATION_KIND_NONE, should this block also be in `pre`? Or should this only run when going from ground to air / vice versa (to halt momentum)?``

Comment (jobrien97) [reply fighters/koopa/src/status.rs]: Sorry, I may not have phrased it clearly; you dont need to wrap the block in an `is_situation_changed` check in the pre-main loop function. It's only necessary for the main loop.

So the point is to be changing motion/kinetic type based on situation *once* in `special_n_main` (once you enter the status), then checking for `is_situation_changed` every frame in `special_n_main_loop` (covering a change in situation in the middle of the move).

Comment (jobrien97) [reply fighters/koopa/src/status.rs]: So put the same code block in both main and main_loop, but wrap the block in `is_situation_changed` in main_loop

**Review by jobrien97 — DISMISSED** [review]
fixa da merge conflicts and we good

**Review by jobrien97 — DISMISSED** [review]
Let's Get This Merged

**Review by SkewedAskew — APPROVED** [review]

**Pull request merged**